cmd/skopeo/copy.go

```diff
@@ -1,71 +1,40 @@
 package main
 
 import (
-	"encoding/json"
-
 	"github.com/Sirupsen/logrus"
 	"github.com/codegangsta/cli"
-	"github.com/projectatomic/skopeo/docker/utils"
+	"github.com/projectatomic/skopeo/image"
 	"github.com/projectatomic/skopeo/signature"
 )
 
-// FIXME: Also handle schema2, and put this elsewhere:
-// docker.go contains similar code, more sophisticated
-// (at the very least the deduplication should be reused from there).
-func manifestLayers(manifest []byte) ([]string, error) {
-	man := manifestSchema1{}
-	if err := json.Unmarshal(manifest, &man); err != nil {
-		return nil, err
-	}
-
-	layers := []string{}
-	for _, layer := range man.FSLayers {
-		layers = append(layers, layer.BlobSum)
-	}
-	return layers, nil
-}
-
-// FIXME: this is a copy from docker_image.go and does not belong here.
-type manifestSchema1 struct {
-	Name     string
-	Tag      string
-	FSLayers []struct {
-		BlobSum string `json:"blobSum"`
-	} `json:"fsLayers"`
-	History []struct {
-		V1Compatibility string `json:"v1Compatibility"`
-	} `json:"history"`
-	// TODO(runcom) verify the downloaded manifest
-	//Signature []byte `json:"signature"`
-}
-
 func copyHandler(context *cli.Context) {
 	if len(context.Args()) != 2 {
 		logrus.Fatal("Usage: copy source destination")
 	}
 
-	src, err := parseImageSource(context, context.Args()[0])
+	rawSource, err := parseImageSource(context, context.Args()[0])
 	if err != nil {
 		logrus.Fatalf("Error initializing %s: %s", context.Args()[0], err.Error())
 	}
+	src := image.FromSource(rawSource)
 
 	dest, err := parseImageDestination(context, context.Args()[1])
 	if err != nil {
 		logrus.Fatalf("Error initializing %s: %s", context.Args()[1], err.Error())
 	}
 	signBy := context.String("sign-by")
 
-	manifest, _, err := src.GetManifest([]string{utils.DockerV2Schema1MIMEType})
+	manifest, err := src.Manifest()
```

Comment:
We don’t, here.

AFAICS there aren’t any cases when the images has both v2s1 and v2s2 and we would want to use the v2s1 one; the “right” one (and also the one that matches a digest reference) is v2s2. Is that incorrect?

For signature verification to make sense, we really need a `types.Image` to be a _single_ image; not a set of several possibly completely different images, with different methods and different parameters to these methods choosing any of them.

So I expect that `types.Image` would, for example:
- Prefer OCI to v2s2, and v2s2 to v2s1
- Given a single `ImageSource` with the same contents, always choose the same version.

If we really needed to refer to the v2s1 manifest of a combined v2s1+v2s2 `ImageSource`, that would be a different `types.Image` object. (`image.FromSourceWithSpecifiedMIMEType`?) and very rarely used.